crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,909 @@
+        let after_index_slice_len = chunk_capacity - index;
+        let len = after_index_slice_len.min(ideal_len) as usize;
+
+        // Relaxed is fine since caller ensures we are iitialized already.
+        // In order for the caller to guarantee that, they must have an ordering that orders this get after the required `set`.
+        let head = self.first.load(Ordering::Relaxed);
+
+        // SAFETY: Caller ensures we are init, so the chunk was allocated via a `Vec` and the index is within the capacity.
+        unsafe { core::slice::from_raw_parts(head, len) }
+    }
+
+    /// Sets this entity at this index.
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    /// Index must be in bounds.
+    /// Access does not conflict with another [`Self::get`].
+    #[inline]
+    unsafe fn set(&self, index: u32, entity: Entity, chunk_capacity: u32) {
+        // Relaxed is fine here since this is not called concurrently and does not conflict with a `get`.
+        let ptr = self.first.load(Ordering::Relaxed);
+        let head = if ptr.is_null() {
+            self.init(chunk_capacity)
+        } else {
+            ptr
+        };
+
+        // SAFETY: caller ensures it is in bounds and we are not fighting with other `set` calls or `get` calls.
+        // A race condition is therefore impossible.
+        let target = unsafe { &*head.add(index as usize) };
+
+        target.set_entity(entity);
+    }
+
+    /// Initializes the chunk to be valid, returning the pointer.
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    #[cold]
+    unsafe fn init(&self, chunk_capacity: u32) -> *mut Slot {
+        let mut buff = ManuallyDrop::new(Vec::new());
+        buff.reserve_exact(chunk_capacity as usize);
+        buff.resize_with(chunk_capacity as usize, Slot::empty);
+        let ptr = buff.as_mut_ptr();
+        // Relaxed is fine here since this is not called concurrently.
+        self.first.store(ptr, Ordering::Relaxed);
+        ptr
+    }
+
+    /// Frees memory
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    /// `chunk_capacity` must be the same as it was initialized with.
+    unsafe fn dealloc(&self, chunk_capacity: u32) {
+        // Relaxed is fine here since this is not called concurrently.
+        let to_drop = self.first.load(Ordering::Relaxed);
+        if !to_drop.is_null() {
+            // SAFETY: This was created in [`Self::init`] from a standard Vec.
+            unsafe {
+                Vec::from_raw_parts(to_drop, chunk_capacity as usize, chunk_capacity as usize);
+            }
+        }
+    }
+}
+
+/// This is a buffer that has been split into chunks, so that each chunk is pinned in memory.
+/// Conceptually, each chunk is put end-to-end to form the buffer.
+/// This will expand in capacity as needed, but a separate system must track the length of the list in the buffer.
```

Comment:
```suggestion
/// This is a buffer that has been split into power-of-two sized chunks, so that each chunk is pinned in memory.
/// Conceptually, each chunk is put end-to-end to form the buffer. This ultimately avoids copying elements on resize,
/// while allowing it to expand in capacity as needed. A separate system must track the length of the list in the buffer.
/// Each chunk is twice as large as the last, except for the first two which have a capacity of 256.
```

Reply:
```suggestion
/// This is a buffer that has been split into power-of-two sized chunks, so that each chunk is pinned in memory.
/// Conceptually, each chunk is put end-to-end to form the buffer. This ultimately avoids copying elements on resize,
/// while allowing it to expand in capacity as needed. A separate system must track the length of the list in the buffer.
/// Each chunk is twice as large as the last, except for the first two which have a capacity of 512.
``` 
I like your wording better. It's 512 though, not 256. Is there an incorrect comment you got that from? I need to fix that if so.